README.md
# [Toast Editor](http://toastedware.com)

Toast Editor is a fast web-based HTML5 level editor that exports your
creations for multiple languages, game frameworks and editors including FlashPunk, Toast, C++, AS3 and Ogmo Editor. 


It's a great choice for fast prototyping or during game jams: it exports
your creation directly in your favorite API, without your code even knowing about
a specific XML format.
Then, if you need it, you can further customize your level with Ogmo Editor using the simple
export feature.

This project has been created by me ([Danny Calleri](http//www.twitter.com/dannycalleri)) as
the perfect companion for [Toast Framework](https://github.com/dannycalleri/toast), a simple C++ 2D game framework.

* Source: [https://github.com/dannycalleri/toasteditor](https://github.com/dannycalleri/toasteditor)
* Homepage: [http://toastedware.com](http://toastedware.com)
* Twitter: [@dannycalleri](http//www.twitter.com/dannycalleri)


## Quick start

Choose one of the following options:

1. Check out the latest stable release from
   [toastedware.com](http://toastedware.com/).
2. Clone the git repo — `git clone
   https://github.com/dannycalleri/toasteditor.git` - and checkout the tagged
   release you'd like to use.


## Features

* Works on almost every platform running a web browser.
* User-friendly 3-steps workflow: load your tileset, create your level and export.
* Retractable and draggable widgets to help you focus on your creation.
* Exports ready-to-use code for FlashPunk and Toast frameworks.
* Exports generic array level maps for C++ or AS3.
* Exports project and level files for Ogmo Editor to further customize your levels.
* [HTML5 Boilerplate](https://github.com/h5bp/html5-boilerplate) guarantees 
  cross-browser compatibility.
* The latest [jQuery](http://jquery.com/) version available thanks to 
  [HTML5 Boilerplate](https://github.com/h5bp/html5-boilerplate).


## Contributing

Anyone and everyone is welcome to [contribute](CONTRIBUTING.md). 


CHANGELOG.md
### HEAD

### 0.2 (October 16, 2013)

* Added Ogmo Editor export option

### 0.1 (October 8, 2013)

* Added export support for FlashPunk, Toast Framework
* Added export support for C++/AS3 using a simple array export
* Tools updated with eraser
* Windows can now be dragged using jQuery UI

js/main.js
jQuery.fn.centerFixed = function () {
    this.css("position","fixed");
    this.css("top", Math.max(0, (($(window).height() - $(this).outerHeight()) / 2) + 
                                                $(window).scrollTop()) + "px");
    this.css("left", Math.max(0, (($(window).width() - $(this).outerWidth()) / 2) + 
                                                $(window).scrollLeft()) + "px");
    return this;
}

jQuery.fn.center = function () {
    this.css("position","absolute");
    this.css("top", Math.max(0, (($(window).height() - $(this).outerHeight()) / 2) + 
                                                $(window).scrollTop()) + "px");
    this.css("left", Math.max(0, (($(window).width() - $(this).outerWidth()) / 2) + 
                                                $(window).scrollLeft()) + "px");
    return this;
}

var closeWindow = function() {
    $('.mouth').addClass('closed');
    $('.window-left-toast').addClass('closed');
    $('.window-right-toast').addClass('closed');
    $('.window-center-col').addClass('closed');
    $('.window-container').hide();
};

var openWindow = function() {
    $('.mouth').removeClass('closed');
    $('.window-left-toast').removeClass('closed');
    $('.window-right-toast').removeClass('closed');
    $('.window-center-col').removeClass('closed');
    $('.window-container').show();
};

var toggleWindow = function() {
    $('.mouth').toggleClass('closed');
    $('.window-left-toast').toggleClass('closed');
    $('.window-right-toast').toggleClass('closed');
    $('.window-center-col').toggleClass('closed');
    $('.window-container').toggle();
};

var resetTools = function() {
    $('#pencil').removeClass('selected');
    $('#eraser').removeClass('selected');
    $('#pencil').removeClass('inactive');
    $('#eraser').removeClass('inactive');
};

var selectPencil = function() {
    isEraserSelected=false;
    resetTools();
    $('#pencil').addClass('selected');
    $('#eraser').addClass('inactive');
    openWindow();

    $("#eraser-selected-balloon").hide();
};

var selectEraser = function() {
    isEraserSelected=true;
    resetTools();
    $('#eraser').addClass('selected');
    $('#pencil').addClass('inactive');
    closeWindow();

    $("#eraser-selected-balloon").show();
    $("#eraser-selected-balloon").css("left", $('.window').width()/2 + 30);
};

//initialize selectbox
$('select').selectBox();
// init
$('#map-container').hide();
$('#tileset-widget').hide();
$('#export-widget').hide();
$('#tools-widget').hide();
$('#data-widget').center();
$('#export-widget').centerFixed(); 
$('#data-form').validate();
$('#new-button').addClass("disabled");
$('#export-button').addClass("disabled");

var initWidth = 229;
console.log("INIT WIDTH = "+initWidth);

// pencil is selected by default
selectPencil();

$('#pencil').click(function(){
    selectPencil();
});

$('#eraser').click(function(){
    selectEraser();
});

$('#tileset-widget-button').click(function(){
    if(isEraserSelected)
    {
        selectPencil();
        toggleWindow();
    }
});

window.onresize = function(event) {
    $('#map-container').center();
    $('#data-widget').center();
    $('#export-widget').centerFixed(); 

    if($('#map-container').offset().top < 52)
    {
        $('#map-container').css('top', '72px');
    }

    if($('#data-widget').offset().top < 52)
    {
        $('#data-widget').css('top', '72px');
    }

    if($('#export-widget').offset().top < 52)
    {
        $('#export-widget').css('top', '72px');
    }
};

var setupWindows = function() {
    // leave 30px for the scrollbar
    // and 20*2px for the toast background
    $('.window-container').width($('.window-container').width() + 30);
    $('.window-center-col').width($('.window-container').width());
    $('.window').width($('.window-container').width() + 20*2);

    $('.window-header-button').click(function(event) {

        if(!isEraserSelected)
        {
            toggleWindow();
        }
    });

    // jQuery UI for dragging
    $('.window').draggable();
};

$('#language-select').change(function() {
    console.log('select changed = '+$("#language-select").val());
    languageChanged($("#language-select").val());
});

var languageChanged = function(language) {

    if(language === "cpp")
    {
        $('#export-text-area').show();
        $('#ogmo-project-export').hide();
        $('#ogmo-level-export').hide();
        $('#export-text-area').val(printLevelCpp());
    }
    else if(language === "as3")
    {
        $('#export-text-area').show();
        $('#ogmo-project-export').hide();
        $('#ogmo-level-export').hide();
        $('#export-text-area').val(printLevelAS3());
    }
    else if(language === "flashpunk")
    {
        $('#export-text-area').show();
        $('#ogmo-project-export').hide();
        $('#ogmo-level-export').hide();
        $('#export-text-area').val(printLevelFlashPunk());
    }
    else if(language === "toast")
    {
        $('#export-text-area').show();
        $('#ogmo-project-export').hide();
        $('#ogmo-level-export').hide();
        $('#export-text-area').val(printLevelToast());
    }
    else if(language === "ogmo")
    {
        $('#export-text-area').hide();
        $('#ogmo-project-export').show();
        $('#ogmo-level-export').show();

        $('#ogmo-project-area').val(printOgmoProject());
        $('#ogmo-level-area').val(printOgmoLevel());
    }
};



js/tinytoast.js


// global variables

// source point from the tiles canvas
// in the format (col, row)
var srcPoint = new Point(0,0);
// point to draw the square in the
// tileset
var squarePoint = new Point(0,0);
var tileW = 0;
var tileH = 0;
var tilesetArray = [];
var editorTilesPerRow = 8;
var emptyChar = 0;

/* 
	bool to check if currently selected tile is transparent:
   	this feature has been discontinued because of cross-origin
   	data problems
*/
//var isSelectedTileTransparent=false;

// tells if a tile has
// already been selected
var isTileSelected=false;

var isEraserSelected=false;

// array that stores the current level
var levelTiles = [];

function printOgmoProject()
{
	var string = "";

	string += "<?xml version=\"1.0\"?>\n";
	string += "<project xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\">\n";
	string += "<OgmoVersion>2.1.0.6</OgmoVersion>\n";
	string += "<Name>New Project</Name>\n";
	string += "<BackgroundColor A=\"255\" R=\"125\" G=\"160\" B=\"255\" />\n";
	string += "<GridColor A=\"255\" R=\"255\" G=\"240\" B=\"90\" />\n";
	string += "<LevelDefaultSize>\n";
	string += "<Width>"+ levelTiles[0].length * tileW +"</Width>\n";
	string += "<Height>"+ levelTiles.length * tileH +"</Height>\n";
	string += "</LevelDefaultSize>\n";
	string += "<LevelMinimumSize>\n";
	string += "<Width>"+ levelTiles[0].length * tileW +"</Width>\n";
	string += "<Height>"+ levelTiles.length * tileH +"</Height>\n";
	string += "</LevelMinimumSize>\n";
	string += "<LevelMaximumSize>\n";
	string += "<Width>"+ levelTiles[0].length * tileW +"</Width>\n";
	string += "<Height>"+ levelTiles.length * tileH +"</Height>\n";
	string += "</LevelMaximumSize>\n";
	string += "<Filename>project.oep</Filename>\n";
	string += "<AngleMode>Radians</AngleMode>\n";
	string += "<CameraEnabled>false</CameraEnabled>\n";
	string += "<CameraSize>\n";
	string += "<Width>"+ levelTiles[0].length * tileW +"</Width>\n";
	string += "<Height>"+ levelTiles.length * tileH +"</Height>\n";
	string += "</CameraSize>\n";
	string += "<ExportCameraPosition>false</ExportCameraPosition>\n";
	string += "<LevelValueDefinitions />\n";
	string += "<LayerDefinitions>\n";
	string += "<LayerDefinition xsi:type=\"TileLayerDefinition\">\n";
	string += "<Name>Tiles</Name>\n";
	string += "<Grid>\n";
	string += "<Width>"+ tileW +"</Width>\n";
	string += "<Height>"+ tileH +"</Height>\n";
	string += "</Grid>\n";
	string += "<ScrollFactor>\n";
	string += "<X>1</X>\n";
	string += "<Y>1</Y>\n";
	string += "</ScrollFactor>\n";
	string += "<ExportMode>CSV</ExportMode>\n";
	string += "</LayerDefinition>\n";
	string += "</LayerDefinitions>\n";
	string += "<Tilesets>\n";
	string += "<Tileset>\n";
	string += "<Name>Tileset</Name>\n";
	string += "<FilePath>"+tilesetFileName+"</FilePath>\n";
	string += "<TileSize>\n";
	string += "<Width>"+ tileW +"</Width>\n";
	string += "<Height>"+ tileH +"</Height>\n";
	string += "</TileSize>\n";
	string += "<TileSep>0</TileSep>\n";
	string += "</Tileset>\n";
	string += "</Tilesets>\n";
	string += "<EntityDefinitions />\n";
	string += "</project>\n";

	return string;
}

function printOgmoLevel()
{
	var string = "";

	string += "<level width=\""+ levelTiles[0].length * tileW +"\" height=\""+ levelTiles.length * tileH +"\">\n";
	string += "<Tiles tileset=\"Tileset\" exportMode=\"CSV\">";

	for(var i=0; i < levelTiles.length; i++)
	{
		//string += "[";

		for(var j=0; j < levelTiles[i].length; j++)
		{
			string += levelTiles[i][j];

			if(j < levelTiles[i].length - 1)
				string += ", ";
		}

		if(i < levelTiles.length - 1)
			string += "\n";
	}

	string += "</Tiles>\n";
	string += "</level>\n";

	return string;
}

function printLevelCpp()
{
	var string = "";

	string += "// An awesome array ready to be loaded in your code \\m/ \n\n";

	string += "int level["+levelTiles.length+"]["+levelTiles[0].length+"] = \n";
	string += "{\n";

	for(var i=0; i < levelTiles.length; i++)
	{
		string += "{";

		for(var j=0; j < levelTiles[i].length; j++)
		{
			string += levelTiles[i][j];

			if(j < levelTiles[i].length - 1)
				string += ", ";
		}

		if(i < levelTiles.length - 1)
			string += "}, \n";
		else
			string += "} \n";
	}

	string += "};";

	return string;
}

function printLevelAS3()
{
	var string = "";

	string += "// An awesome array ready to be loaded in your code \\m/ \n\n";

	string += "var level:Array = \n";
	string += "[\n";

	for(var i=0; i < levelTiles.length; i++)
	{
		string += "[";

		for(var j=0; j < levelTiles[i].length; j++)
		{
			string += levelTiles[i][j];

			if(j < levelTiles[i].length - 1)
				string += ", ";
		}

		if(i < levelTiles.length - 1)
			string += "], \n";
		else
			string += "] \n";
	}

	string += "];";

	return string;

	return string;
}

function printLevelToast()
{
	var string = "";

	string += "// Add this code to an Entity's subclass (e.g. in the constructor) \n\n";

	string += "std::string level = ";
	for(var i=0; i < levelTiles.length; i++)
	{
		string += "\"";

		for(var j=0; j < levelTiles[i].length; j++)
		{
			string += levelTiles[i][j];

			if(j < levelTiles[i].length - 1)
				string += ", ";
		}

		if(i < levelTiles.length - 1)
			string += "\\n\" \n";
		else
			string += "\\n\"";
	}

	string += "; \n\n";

	string += "Tilemap* tilemap = new Tilemap(\""+tilesetFileName+"\", "+
				levelTiles[0].length * tileW + ", " + 
				levelTiles.length * tileH + ", " + tileW+", "+tileH+");" + "\n";

	string += "tilemap->loadFromString(level);\n"
	string += "graphic = tilemap;\n";

	return string;
}

function printLevelFlashPunk()
{
	var string = "";

	string += "var level:String = ";
	for(var i=0; i < levelTiles.length; i++)
	{
		string += "\"";
		
		for(var j=0; j < levelTiles[i].length; j++)
		{
			string += levelTiles[i][j];

			if(j < levelTiles[i].length - 1)
				string += ",";
		}

		if(i < levelTiles.length - 1)
			string += "\\n\"+\n";
	}

	string += "\"\n\n";

	string += "var tilemap:Tilemap = new Tilemap(\""+tilesetFileName+"\", "+
				levelTiles[0].length * tileW + ", " + 
				levelTiles.length * tileH + ", " + tileW+", "+tileH+");" + "\n\n";

	string += "tilemap.loadFromString(level, \",\", \"\\n\");";

	return string;
}

// little Point class helper
function Point(x, y)
{
	this.x = x;
	this.y = y;
}

// global helper methods
var getMousePos = function (canvas, evt) {
	var rect = canvas.getBoundingClientRect();
	return [evt.clientX - rect.left, evt.clientY - rect.top];
};

// mouse helpers
var mouseDown = 0;
document.body.onmousedown = function() { 
  ++mouseDown;
};

document.body.onmouseup = function() {
  --mouseDown;
};


function clearCanvas(canvas, color) 
{
	if(canvas.getContext) 
	{
		// Initaliase a 2-dimensional drawing context
		var context = canvas.getContext('2d');
		//Canvas commands go here

		context.fillStyle = color;
		context.fillRect(0, 0, canvas.width, canvas.height);
	}
}

function clearCanvasTransparent(canvas) 
{
	if(canvas.getContext) 
	{
		var context = canvas.getContext('2d');
		context.clearRect(0, 0, canvas.width, canvas.height);
	}
}


function ToastCanvas()
{
	this.rows = 0;
	this.cols = 0;
	this.tileWidth = 0;
	this.tileHeight = 0;
	this.width = 0;
	this.height = 0;
}

/*
*	Canvas for the map	
*/
function MapCanvas(cols, rows, tileWidth, tileHeight)
{
	this.rows = rows;
	this.cols = cols;
	tileW = tileWidth;
	tileH = tileHeight;
	this.tileWidth = tileWidth;
	this.tileHeight = tileHeight;

	// local width and height
	var width = cols * tileWidth;
	var height = rows * tileHeight;
	// public variables
	this.width = width;
	this.height = height;

	// public methods
	//this.clearCanvas = clearCanvas;
	this.resizeCanvas = resizeCanvas;
	this.drawGrid = drawGrid;

	var mapCanvas = document.getElementById('mapCanvas');

	// mapCanvas clicked
	mapCanvas.addEventListener('mousedown', function(evt) {

		if(isTileSelected)
		{
			mouseDown=0;
			var mousePos = getMousePos(mapCanvas, evt);
			var col = Math.floor(mousePos[0] / tileWidth);
			var row = Math.floor(mousePos[1] / tileHeight);

			updateLevel(col, row);

			if(mapCanvas.getContext) 
			{
				var img = document.getElementById("tileset");
				var ctx = mapCanvas.getContext("2d");

				if(!isEraserSelected)
				{
					ctx.fillRect(col*tileW, row*tileH, tileW, tileH);
					ctx.drawImage(img, srcPoint.x * tileW, srcPoint.y * tileH, tileW, tileH, col * tileW, row * tileH, tileW, tileH);
				}
				else
				{
					ctx.fillRect(col*tileW, row*tileH, tileW, tileH);
				}

				//drawGrid();
			}
		}

	}, false);
	
	mapCanvas.addEventListener('mousemove', function(evt) {

		if(mouseDown && isTileSelected)
		{
			var mousePos = getMousePos(mapCanvas, evt);
			var col = Math.floor(mousePos[0] / tileWidth);
			var row = Math.floor(mousePos[1] / tileHeight);

			updateLevel(col, row);

			if(mapCanvas.getContext) 
			{
				var ctx = mapCanvas.getContext("2d");
				var img = document.getElementById("tileset");

				if(!isEraserSelected)
				{
					ctx.fillRect(col*tileW, row*tileH, tileW, tileH);
					ctx.drawImage(img, srcPoint.x * tileW, srcPoint.y * tileH, tileW, tileH, col * tileW, row * tileH, tileW, tileH);
				}
				else
				{
					ctx.fillRect(col*tileW, row*tileH, tileW, tileH);
				}

				//drawGrid();
			}
			
		}

	}, false);

	// redraw grid when mouse is up
	mapCanvas.addEventListener('mouseup', function(evt) {
		drawGrid();
	}, false);
	

	var defaultColor = "#000";

	// gives the canvas the new size
	// and creates the array to store
	// the level
	resizeCanvas(this.width, this.height);
	drawGrid();

	function updateLevel(col, row)
	{
		var img = document.getElementById("tileset");
		// storing value inside level array
		var tilesetCols = (img.width / tileW);

		if(!isEraserSelected)
		{
			levelTiles[row][col] = (srcPoint.y * tilesetCols) + srcPoint.x;
		}
		else
		{
			levelTiles[row][col] = emptyChar;
		}

		/*
		console.log("\n");
		for(var i=0; i < levelTiles.length; i++)
		{
			console.log(levelTiles[i]);
		}
		*/
		
	}

	function resizeCanvas(width, height)
	{
		mapCanvas.width = width;
		mapCanvas.height = height;
		clearCanvas(mapCanvas, defaultColor);

		// creating the level map
		levelTiles = new Array(rows);
		for(var i=0; i < rows; i++)
		{
			levelTiles[i] = new Array(cols);
		}

		for(var i=0; i < rows; i++)
		{
			for(var j=0; j < cols; j++)
			{
				levelTiles[i][j] = emptyChar;
			}
		}
	}

	function drawGrid()
	{
		if(mapCanvas.getContext) 
		{
			var ctx = mapCanvas.getContext('2d');
			ctx.lineWidth = 1;
			ctx.strokeStyle="#333";

			// drawing with 0.5 offset
			// for pixel perfect lines
			// (OpenGL style)
			for(var i=1; i < rows; i++)
			{
				ctx.moveTo(0.5, i*tileHeight + 0.5);
				ctx.lineTo(width + 0.5, i*tileHeight + 0.5);
				ctx.stroke();
			}


			for(var i=1; i < cols; i++)
			{
				ctx.moveTo(i*tileWidth + 0.5, 0.5);
				ctx.lineTo(i*tileWidth + 0.5, height + 0.5);
				ctx.stroke();
			}
			
		}
	}

	/*
	function getMousePos(canvas, evt) 
	{
		var rect = canvas.getBoundingClientRect();
		return [evt.clientX - rect.left, evt.clientY - rect.top];
	}
	*/
}

MapCanvas.prototype = new ToastCanvas();

function TilesCanvas()
{
	var clearColor = "#eee";
	var canvas = document.getElementById('tilesCanvas');
	//drawTileset();

	var img = document.getElementById("tileset");
	//img.crossOrigin="";
	/*
	var src = "img/tileset.png";
	img.crossOrigin = "Anonymous";
	img.onload = function() {
	    canvas.width = img.width;
	    canvas.height = img.height;

	    if(canvas.getContext)
	    {
	    	var ctx = canvas.getContext("2d");
	    	ctx.drawImage( img, 0, 0 );
		}

	    localStorage.setItem( "savedImageData", canvas.toDataURL("image/png") );
	}
	img.src = src;
	// make sure the load event fires for cached images too
	if ( img.complete || img.complete === undefined ) {
	    img.src = "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///ywAAAAAAQABAAACAUwAOw==";
	    img.src = src;
	}
	*/

	// read the entire tileset, create Tile objects and save them in an array
	readTileset();
	// draw Tile objects in a the canvas one by one
	drawTileset();

	// create the algorithm for 

	// SCALE CANVAS
	/*
	if(canvas.getContext)
	{
		// scale y component
	    //ctx.scale(0.5, 0.5);
	}
	*/

	/*
	var testimg = new Image,
	    imgCanvas = document.createElement("canvas"),
	    ctx = imgCanvas.getContext("2d"),
	    src = "../img/tileset.png"; // insert image url here

	testimg.crossOrigin = "Anonymous";

	testimg.onload = function() {
	    imgCanvas.width = testimg.width;
	    imgCanvas.height = testimg.height;
	    ctx.drawImage( testimg, 0, 0 );
	    localStorage.setItem( "savedImageData", imgCanvas.toDataURL("image/png") );
	}
	testimg.src = src;
	// make sure the load event fires for cached images too
	if ( testimg.complete || testimg.complete === undefined ) 
	{
	    testimg.src = "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///ywAAAAAAQABAAACAUwAOw==";
	    testimg.src = src;
	}
	*/
	
	canvas.addEventListener('click', function(evt) {

		isTileSelected=true;
		//isSelectedTileTransparent=false;
		var mousePos = getMousePos(canvas, evt);
		var col = Math.floor(mousePos[0] / tileW);
		var row = Math.floor(mousePos[1] / tileH);
		// updating src point
		// to render on map canvas

		/*
		srcPoint.x = col;
		srcPoint.y = row;
		*/

		var tilesetIndex = row * editorTilesPerRow + col;
		// src point (col,row) from which to draw
		var point = tilesetArray[tilesetIndex];
		srcPoint.x = point.x;
		srcPoint.y = point.y;

		squarePoint.x = col;
		squarePoint.y = row;

		var centerPointX = (squarePoint.x * tileW) + tileW/2;
		var centerPointY = (squarePoint.y * tileH) + tileH/2;
		
		// check to see if selected tile
		// is transparent
		if(canvas.getContext) 
		{
			// WARNING - this code may explode!
			// Please leave the following code commented.
			// It works only when the editor
			// is running from a server
			// WARNING
			/*
			var ctx = canvas.getContext("2d");
			var imageData = ctx.getImageData(centerPointX, centerPointY, 1, 1).data;
			console.log("ALPHA = "+imageData[3]);

			// if pixel is transparent
			// (thus the tile is transparent)
			if(imageData[3] == 0)
			{
				isSelectedTileTransparent=true;
			}
			*/
		}
		

		drawSquare();

		// debug
		/*
		if(canvas.getContext) 
		{
			var ctx = canvas.getContext("2d");
			ctx.fillStyle="#FF0000";
			ctx.fillRect(centerPointX, centerPointY, 5, 5);
		}
		*/

		//var message = '[TilesCanvas] Mouse position: ' + col + ',' + row;
		//console.log(message);

	}, false);
	

	function readTileset()
	{
		if(canvas.getContext) 
		{
			var ctx = canvas.getContext("2d");

			var cols = Math.floor(img.width / tileW);
			var rows = Math.floor(img.height / tileH);

			for(var row=0; row < rows; row++)
			{
				for(var col=0; col < cols; col++)
				{
					var point = new Point(0,0);
					point.x = col;
					point.y = row;
					tilesetArray.push(point);
				}
			}

			// debug
			/*
			for(var i=0; i < tilesetArray.length; i++)
			{
				var point = tilesetArray[i];
				console.log(point);
			}
			*/
		}

		canvas.width = editorTilesPerRow * tileW;
		canvas.height = Math.floor(tilesetArray.length / editorTilesPerRow) * tileH;
		clearCanvasTransparent(canvas);

		// reset previous style
		ctx.lineWidth = 3;
		ctx.strokeStyle = "#ffff00";
	}

	function drawTileset()
	{
		// draw Tile objects in a the canvas one by one
		for(var i=0; i < tilesetArray.length; i++)
		{
			var point = tilesetArray[i];
			// % numTilesPerRow
			var destCol = (i % editorTilesPerRow);
			var destRow = Math.floor(i / editorTilesPerRow);

			//console.log(destCol + " " + destRow);

			if(canvas.getContext) 
			{
				var ctx = canvas.getContext("2d");
				var img = document.getElementById("tileset");
				ctx.drawImage(img, point.x * tileW, point.y * tileH, tileW, tileH, destCol * tileW, destRow * tileH, tileW, tileH);
			}
		}
	}

	function drawSquare()
	{
		if(canvas.getContext)
		{
			var ctx = canvas.getContext("2d");
			//ctx.restore();
			ctx.lineWidth = 3;
			ctx.strokeStyle = "#ffff00";

			clearCanvasTransparent(canvas);
			drawTileset();
			ctx.strokeRect(squarePoint.x * tileW + 0.5, squarePoint.y * tileH + 0.5, tileW, tileH);
		}
	}

	/*
	function drawTileset()	
	{
		if(canvas.getContext) 
		{
			var ctx = canvas.getContext("2d");
			var img = document.getElementById("tileset");

			//ctx.save();
			
			canvas.width = img.width;
			canvas.height = img.height;
			clearCanvas(canvas, clearColor);

			// reset previous style
			ctx.lineWidth = 3;
			ctx.strokeStyle = "#ffff00";

			ctx.drawImage(img, 0, 0);
		}
	}
	*/

}

TilesCanvas.prototype = new ToastCanvas();




















js/dragdroploader.js
// The following code is taken from a great blog post
// by Robert Gravelle. You can find it here:
// URL = http://www.htmlgoodies.com/html5/javascript/drag-files-into-the-browser-from-the-desktop-HTML5.html

var tilesetLoaded = false;
var tilesetFileName = "";

if(window.FileReader)
{
    var drop;
    
    addEventHandler(window, 'load', function() {
        
        var status = document.getElementById('status');
        drop = document.getElementById('drop');
        var list = document.getElementById('list');

        function cancel(e) 
        {
            if (e.preventDefault) { e.preventDefault(); }
            return false;
        }

        // Tells the browser that we *can* drop on this target
        addEventHandler(drop, 'dragover', cancel);
        addEventHandler(drop, 'dragenter', cancel);

        addEventHandler(drop, 'drop', function (e) {
            e = e || window.event; // get window.event if e argument missing (in IE)   
            if (e.preventDefault) { e.preventDefault(); } // stops the browser from redirecting off to the image.

            var dt    = e.dataTransfer;
            var files = dt.files;

            for (var i=0; i < files.length; i++)
            {
                var file = files[i];
                var reader = new FileReader();

                //attach event handlers here...

                reader.readAsDataURL(file);
                
                addEventHandler(reader, 'loadend', function(e, file) {
                    var bin           = this.result; 
                    var newFile       = document.createElement('div');
                    newFile.innerHTML = 'Loaded : '+file.name+' size '+file.size+' B';
                    tilesetFileName = file.name;

                    list.appendChild(newFile);  
                    
                    var fileNumber = list.getElementsByTagName('div').length;
                    status.innerHTML = fileNumber < files.length 
                                 ? 'Loaded 100% of file '+fileNumber+' of '+files.length+'...' 
                                 : 'Done loading. processed '+fileNumber+' files.';

                    var img = document.createElement("img"); 
                    img.setAttribute("id", "tileset");
                    img.file = file;
                    img.src = bin;
                    document.getElementById("tileset-container").appendChild(img);

                    tilesetLoaded=true;

                    //list.appendChild(img);
                }.bindToEventHandler(file));
            }
            
            return false;
        });

        Function.prototype.bindToEventHandler = function bindToEventHandler() 
        {
            var handler = this;
            var boundParameters = Array.prototype.slice.call(arguments);
            //create closure
            return function(e) 
            {
                e = e || window.event; // get window.event if e argument missing (in IE)   
                boundParameters.unshift(e);
                handler.apply(this, boundParameters);
            }
        };
    });
} 
else 
{ 
    document.getElementById('status').innerHTML = 'Your browser does not support the HTML5 FileReader.';
}

function addEventHandler(obj, evt, handler) 
{
    if(obj.addEventListener) 
    {
        // W3C method
        obj.addEventListener(evt, handler, false);
    } 
    else if(obj.attachEvent) 
    {
        // IE method.
        obj.attachEvent('on'+evt, handler);
    } 
    else 
    {
        // Old school method.
        obj['on'+evt] = handler;
    }
}

js/main copia.js

function ToastCanvas()
{
	this.rows = 0;
	this.cols = 0;
	this.tileWidth = 0;
	this.tileHeight = 0;
	this.width = 0;
	this.height = 0;
	this.coppola = "di badda";

	this.greet = greet;

	function greet() {
        console.log("Hi, I'm " + this.coppola);
    };

    this.getMousePos = function (canvas, evt) {
		var rect = canvas.getBoundingClientRect();
		return [evt.clientX - rect.left, evt.clientY - rect.top];
	};
}

/*
ToastCanvas.prototype.getMousePos = function (canvas, evt) 
{
	var rect = canvas.getBoundingClientRect();
	return [evt.clientX - rect.left, evt.clientY - rect.top];
};
*/

/*
ToastCanvas.prototype = {
	getMousePos: function(canvas, evt) {
		var rect = canvas.getBoundingClientRect();
		return [evt.clientX - rect.left, evt.clientY - rect.top];
	}
};
*/

/*
*	Canvas for the map	
*/
function MapCanvas(cols, rows, tileWidth, tileHeight)
{
	this.coppola = "di cazzu";
	this.rows = rows;
	this.cols = cols;
	this.tileWidth = tileWidth;
	this.tileHeight = tileHeight;

	// local width and height
	var width = cols * tileWidth;
	var height = rows * tileHeight;
	// public variables
	this.width = width;
	this.height = height;

	// public methods
	this.clearCanvas = clearCanvas;
	this.resizeCanvas = resizeCanvas;
	this.drawGrid = drawGrid;

	var mapCanvas = document.getElementById('mapCanvas');

	mapCanvas.addEventListener('click', function(evt) {

		var mousePos = getMousePos(mapCanvas, evt);
		var col = Math.floor(mousePos[0] / tileWidth);
		var row = Math.floor(mousePos[1] / tileHeight);
		var message = 'Mouse position: ' + col + ',' + row;
		console.log(message);

	}, false);

	var defaultColor = "#000";
	resizeCanvas(this.width, this.height);
	drawGrid();

	function clearCanvas(color) 
	{
		if(mapCanvas.getContext) 
		{
			// Initaliase a 2-dimensional drawing context
			var context = mapCanvas.getContext('2d');
			//Canvas commands go here

			context.fillStyle = color;
			context.fillRect(0, 0, mapCanvas.width, mapCanvas.height);
		}
	}

	function resizeCanvas(width, height)
	{
		mapCanvas.width = width;
		mapCanvas.height = height;
		clearCanvas(defaultColor);
	}

	function drawGrid()
	{
		if(mapCanvas.getContext) 
		{
			var ctx = mapCanvas.getContext('2d');
			ctx.lineWidth = 1;
			ctx.strokeStyle="#333";

			// drawing with 0.5 offset
			// for pixel perfect lines
			// (OpenGL style)
			for(var i=1; i < rows; i++)
			{
				ctx.moveTo(0.5, i*tileHeight + 0.5);
				ctx.lineTo(width + 0.5, i*tileHeight + 0.5);
				ctx.stroke();
			}


			for(var i=1; i < cols; i++)
			{
				ctx.moveTo(i*tileWidth + 0.5, 0.5);
				ctx.lineTo(i*tileWidth + 0.5, height + 0.5);
				ctx.stroke();
			}
			
		}
	}

	/*
	function getMousePos(canvas, evt) 
	{
		var rect = canvas.getBoundingClientRect();
		return [evt.clientX - rect.left, evt.clientY - rect.top];
	}
	*/
}

MapCanvas.prototype = new ToastCanvas();
//MapCanvas.prototype.constructor = MapCanvas;

function TilesCanvas()
{
	var canvas = document.getElementById('tilesCanvas');
	if(canvas.getContext) 
	{
		var ctx = canvas.getContext("2d");
		var img = document.getElementById("tileset");
		ctx.drawImage(img, 0, 0);
	}
}

TilesCanvas.prototype = new ToastCanvas();
//TilesCanvas.prototype.constructor = TilesCanvas;

var mapCanvas = new MapCanvas(20, 15, 32, 32);
var tilesCanvas = new TilesCanvas();

mapCanvas.greet();
tilesCanvas.greet();

//console.log("MAP CANVAS: " + mapCanvas.coppola);
//console.log("TILES CANVAS: " + tilesCanvas.coppola);

/*
var Person = function() {
    this.canTalk = true;
    this.greet = function() {
        if (this.canTalk) {
            console.log("Hi, I'm " + this.name);
        }
    };
};

var Customer = function(name) {
    this.name = name;
};
Customer.prototype = new Person();

var person = new Person();
var customer = new Customer("Danny");

person.greet();
customer.greet();
*/

//mapCanvas.createGrid();



